Source/WebCore/loader/DocumentPrefetcher.cpp

```diff
@@ -0,0 +1,229 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "DocumentPrefetcher.h"
+
+#include "CachedRawResource.h"
+#include "CachedResourceClient.h"
+#include "CachedResourceLoader.h"
+#include "CachedResourceRequest.h"
+#include "Document.h"
+#include "DocumentLoader.h"
+#include "Frame.h"
+#include "FrameLoader.h"
+#include "NetworkLoadMetrics.h"
+#include "ReferrerPolicy.h"
+#include "ResourceRequest.h"
+#include "SecurityOrigin.h"
+#include "SecurityPolicy.h"
+#include <wtf/Box.h>
+
+namespace WebCore {
+
+DocumentPrefetcher::DocumentPrefetcher(FrameLoader& frameLoader)
+    : m_frameLoader(frameLoader)
+{
+}
+
+DocumentPrefetcher::~DocumentPrefetcher()
+{
+    clear();
+}
+
+static bool isPassingSecurityChecks(const URL& url, Document& document)
+{
+    Ref documentOrigin = document.securityOrigin();
+    Ref urlOrigin = SecurityOrigin::create(url);
+    if (!documentOrigin->isSameOriginAs(urlOrigin)) {
+        document.addConsoleMessage(MessageSource::Security, MessageLevel::Error,
+            makeString("Prefetch request denied: not same origin as document"_s));
+        return false;
+    }
+
+    if (!SecurityOrigin::isSecure(url)) {
+        document.addConsoleMessage(MessageSource::Security, MessageLevel::Error,
+            makeString("Prefetch request denied: URL must be secure (HTTPS)"_s));
+        return false;
+    }
+
+    return true;
+}
+
+static ResourceRequest makePrefetchRequest(const URL& url, const Vector<String>& tags, const String& referrerPolicyString, const URL& referrerURL, const Document& document)
+{
+    String urlString = url.string();
+    ResourceRequest request { WTFMove(urlString) };
+    request.setPriority(ResourceLoadPriority::VeryLow);
+
+    if (!tags.isEmpty()) {
+        StringBuilder builder;
+        for (size_t i = 0; i < tags.size(); ++i) {
+            builder.append(tags[i]);
+            if (i < tags.size() - 1)
```

Comment:
nit: it's a tiny bit more efficient to append a comma before the tag if i is nonzero.